### ISSUE TITLE
ttp and audit events sending the same entry repeatedly

### ISSUE THREAD

**Issue by chris-y** [JoshuaSmeda/mimecast_log_collector#5]
I tried your new version with the ttp and audit events enabled, but the most recent log entry for ttp and audit events appears to be continually sent to the syslog server which is just filling it up with thousands of duplicate messages.

(I'm not sure if it is only the most recent entry or all of them being re-sent, looks like it may be all)

**Comment by JoshuaSmeda** [comment]
Are you using a syslog collector (i.e Logstash) or are you just testing by packet capture such as Wireshark? If syslog, which product are you using?

I will review this issue tonight and see if I can come up with a viable solution.

**Comment by chris-y** [comment]
I'm using a syslog collector - Graylog.
It doesn't filter out duplicate events, I have in the past employed a workaround which tells it to delete messages older than a certain time period, but it's not a workaround I particularly like and I'm not sure it will work here as the import is too frequent.

**Comment by JoshuaSmeda** [comment]
Hey @chris-y 

Take a look at branch `issue5` - https://github.com/JoshuaSmeda/mimecast_log_collector/tree/issue5

I've made some changes that would stop sending duplicate logs. I tested using Wireshark and didn't see duplicates coming through - please confirm using Graylog.

I added hashing to the audit & ttp events. Each mimecast event would be MD5 hashed so it's unique. The hash is generated based on the contents of the event. If a event differs slightly, the hash would be different and therefore log to syslog but if a new event is the exact same as before, (a hash would've previously been generated and stored) it wouldn't be sent to syslog.

Side note: I will need to come up with a better, more efficient (memory, IO) way of storing and querying the hashes. I'll release that at a later stage. I will also add functionality to define the retention period for logs to hang around (right now I also do a workaround but if it was built in, it would be better).

Let me know if it works for you?

**Comment by chris-y** [comment]
With a tweak (because I run this as a daemon), yes it does!

Just added a path to the hash_file file as below:
```diff
diff --git a/audit_events.py b/audit_events.py
index 2890a32..ec9d114 100644
--- a/audit_events.py
+++ b/audit_events.py
@@ -48,12 +48,12 @@ def get_audit_events(base_url, access_key, secret_key):
                             lines = log_file.read().splitlines()
                             for line in lines:
                                 hashed_event = (hashlib.md5(line.encode('utf-8')).hexdigest())
-                                with open('hash_file', 'r') as hash_file:
+                                with open(configuration.logging_details['CHK_POINT_DIR'] + 'hash_file', 'r') as hash_file:
                                     if (hashed_event + '\n') in hash_file.readlines():
                                         print("Hash %s already exists" % (hashed_event))
                                     else:
                                         syslogger.info(line)
-                                        append_file('hash_file', hashed_event)
+                                        append_file(configuration.logging_details['CHK_POINT_DIR'] + 'hash_file', hashed_event)

                         log.info('Syslog output completed for file ' + file_name)

diff --git a/ttp_events.py b/ttp_events.py
index c7ef7e9..8b4e6ed 100644
--- a/ttp_events.py
+++ b/ttp_events.py
@@ -49,12 +49,12 @@ def Get_TTPURL_events(base_url, access_key, secret_key):
                             lines = log_file.read().splitlines()
                             for line in lines:
                                 hashed_event = (hashlib.md5(line.encode('utf-8')).hexdigest())
-                                with open('hash_file', 'r') as hash_file:
+                                with open(configuration.logging_details['CHK_POINT_DIR'] + 'hash_file', 'r') as hash_file:
                                     if (hashed_event + '\n') in hash_file.readlines():
                                         print("Hash %s already exists" % (hashed_event))
                                     else:
                                         syslogger.info(line)
-                                        append_file('hash_file', hashed_event)
+                                        append_file(configuration.logging_details['CHK_POINT_DIR'] + 'hash_file', hashed_event)

                         log.info('Syslog output completed for file ' + file_name)
```

**Comment by JoshuaSmeda** [comment]
Okay that's great news, thanks for the diff.

I also run the program as a daemon (SysVinit specifically) and I just declare the `WorkingDirectory` so I never needed to modify the code like you've done.

This fix has been merged #6 with your advised changes

Thanks again @chris-y  for raising this issue and helping out
